Support for MCUBoot's single application slot mode & USB DFU

nvlsianpu opened the pull request:
It is possible that USB DFU class is used by MCUBoot itself while it has enabled CONFIG_SINGLE_APPLICATIO_SLOT. In this mode only application slot is the primary slot.

This patch assumes that if the secondary slot dosn't exist this mean modules should operates on the primary slots.
I avoided to use MCUBoot configuration output in zephyr code (so `CONFIG_SINGLE_APPLICATIO_SLOT`) - which I believe is a one of zephyr design principle.

Also be aware of the bug  #28504

~~PR includes https://github.com/zephyrproject-rtos/zephyr/pull/28519.~~

***building***
`west build -d build/single_usb -b nrf52840dk_nrf52840 ./bootloader/mcuboot/boot/zephyr/ -t menuconfig -- -DDTC_OVERLAY_FILE=./boards/nrf52840_single_slot.overlay`

set
```
CONFIG_SINGLE_APPLICATION_SLOT=y
CONFIG_BOOT_WAIT_FOR_USB_DFU=y
CONFIG_MULTITHREADING=y
CONFIG_SIZE_OPTIMIZATIONS=y
```

`west build -d build/single_usb`

***testing***
Similar as for USB DFU sample need to build for instance the hello_word sample with same DTS overlay file as MCUBoot.

`west build -d build/hellow_single -b nrf52840dk_nrf52840 ./zephyr/samples/hello_world/ -t menuconfig -- -DDTC_OVERLAY_FILE=../../../bootloader/mcuboot/boot/zephyr/boards/nrf52840_single_slot.overlay`

set
`CONFIG_BOOTLOADER_MCUBOOT=y`

`west build -d build/hellow_single`

`west sign -d build/hellow_single/ -t imgtool -- -k ./bootloader/mcuboot/root-rsa-2048.pem`

`west flash -d build/single_usb/`

`sudo dfu-util --alt 0 --download ./build/hellow_single/zephyr/zephyr.signed.bin`

console output:
``` console
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 2fe3:0100
Run-time device DFU version 0110
Claiming USB DFU Runtime Interface...
Determining device status: state = appIDLE, status = 0
Device really in Runtime Mode, send DFU detach request...
Resetting USB...
Opening DFU USB Device...
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 0110
Device returned transfer size 128
Copying data from PC to DFU device
Download        [=========================] 100%        16592 bytes
Download done.
state(2) = dfuIDLE, status(0) = No error condition is present
Done!
```

fixes #28502
fixes #28504

nvlsianpu committed:
dfu: allow to operate on primary slot

It is possible that dfu libraries are used by MCUBoot
itself while it has enabled CONFIG_SINGLE_APPLICATIO_SLOT.
In this mode only application slot is the primary slot.

Patch is required for supports USB DFU class in mentioned
MCUBoot's mode.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
usb/class/usb_dfu: allow to operate only on the primary slot

It is possible that USB DFU class is used by MCUBoot
itself while it has enabled CONFIG_SINGLE_APPLICATIO_SLOT.
In this mode only application slot is the primary slot.

Adapted DFU class descriptor and services to optionally
serve only for one image.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

utzig commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -616,7 +622,7 @@ int mcuboot_swap_type(void)
 	}
 
-	rc = boot_read_swap_state_by_id(FLASH_AREA_IMAGE_SECONDARY,
+	rc = boot_read_swap_state_by_id(UPLOAD_FLASH_AREA_ID,
```
I am not sure this is a good idea, the tables were not build with comparing a partition against itself in mind. I would just check if `UPLOAD_FLASH_AREA_ID` equals `FLASH_AREA_IMAGE_PRIMARY` at the beginning and return `BOOT_SWAP_TYPE_NONE`

utzig commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -586,7 +590,9 @@ boot_read_swap_state_by_id(int flash_area_id, struct boot_swap_state *state)
-	case FLASH_AREA_IMAGE_SECONDARY:
+#if (UPLOAD_FLASH_AREA_ID != FLASH_AREA_IMAGE_PRIMARY)
+	case UPLOAD_FLASH_AREA_ID:
+#endif
```
Could this be?

```
#ifdef FLASH_AREA_IMAGE_SECONDARY
￼	case FLASH_AREA_IMAGE_SECONDARY:￼
#endif
```

utzig commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -75,10 +75,14 @@ struct mcuboot_v1_raw_header {
+	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_1)
+#else
+	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_0)
+#endif
```
I think it would be better to leave `FLASH_AREA_IMAGE_SECONDARY` defined (conditionally) and use *it* when doing swap state checks, and use `UPLOAD_FLASH_AREA_ID` only for operations where sending data to a partition is involved.

nvlsianpu replied:
I agree fully

utzig reviewed (commented):
A few suggestions on the change to check what's your opinion...

nvlsianpu committed:
dfu/boot/mcuboot: made able to compile within MCUBoot

Use rather MCUBoot implementation of some function than
zephyr's implementations once avielabel.

Zephyr implementations are needed while compiling application
supposed to be chainloaded by the MCUBoot or when they are
tested.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

jfischer-phytec-iot commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -298,10 +298,17 @@ struct dfu_data_t {
 	uint16_t bwPollTimeout;
 };
 
+#if FLASH_AREA_LABEL_EXISTS(image_1)
```
This should also be applied to the descriptors

```diff
diff --git a/subsys/usb/class/usb_dfu.c b/subsys/usb/class/usb_dfu.c
index 2c22d4e414..e3a0d3ffb4 100644
--- a/subsys/usb/class/usb_dfu.c
+++ b/subsys/usb/class/usb_dfu.c
@@ -123,7 +123,9 @@ struct dev_dfu_mode_descriptor {
        struct usb_cfg_descriptor cfg_descr;
        struct usb_sec_dfu_config {
                struct usb_if_descriptor if0;
+#if FLASH_AREA_LABEL_EXISTS(image_1)
                struct usb_if_descriptor if1;
+#endif
                struct dfu_runtime_descriptor dfu_descr;
        } __packed sec_dfu_cfg;
 } __packed;
@@ -172,6 +174,7 @@ struct dev_dfu_mode_descriptor dfu_mode_desc = {
                        .bInterfaceProtocol = DFU_MODE_PROTOCOL,
                        .iInterface = 4,
                },
+#if FLASH_AREA_LABEL_EXISTS(image_1)
                .if1 = {
                        .bLength = sizeof(struct usb_if_descriptor),
                        .bDescriptorType = USB_INTERFACE_DESC,
@@ -183,6 +186,7 @@ struct dev_dfu_mode_descriptor dfu_mode_desc = {
                        .bInterfaceProtocol = DFU_MODE_PROTOCOL,
                        .iInterface = 5,
                },
+#endif
                .dfu_descr = {
                        .bLength = sizeof(struct dfu_runtime_descriptor),
                        .bDescriptorType = DFU_FUNC_DESC,
@@ -226,11 +230,13 @@ struct usb_string_desription {
                uint8_t bString[USB_BSTRING_LENGTH(FIRMWARE_IMAGE_0_LABEL)];
        } __packed utf16le_image0;
 
+#if FLASH_AREA_LABEL_EXISTS(image_1)
        struct image_1_descriptor {
                uint8_t bLength;
                uint8_t bDescriptorType;
                uint8_t bString[USB_BSTRING_LENGTH(FIRMWARE_IMAGE_1_LABEL)];
        } __packed utf16le_image1;
+#endif
 } __packed;
 
 USBD_STRING_DESCR_DEFINE(secondary)
@@ -267,6 +273,7 @@ struct usb_string_desription string_descr = {
                .bDescriptorType = USB_STRING_DESC,
                .bString = FIRMWARE_IMAGE_0_LABEL,
        },
+#if FLASH_AREA_LABEL_EXISTS(image_1)
        /* Image 1 String Descriptor */
        .utf16le_image1 = {
                .bLength = USB_STRING_DESCRIPTOR_LENGTH(
@@ -274,6 +281,7 @@ struct usb_string_desription string_descr = {
                .bDescriptorType = USB_STRING_DESC,
                .bString = FIRMWARE_IMAGE_1_LABEL,
        },
+#endif
 };
 
 /* This element marks the end of the entire descriptor. */
```

nvlsianpu replied:
@jfischer-phytec-iot Thanks for this diff, will apply it.

emob-nordic replied:
Add this as well.
``` diff
diff --git a/subsys/usb/class/usb_dfu.c b/subsys/usb/class/usb_dfu.c
index 2c22d4e414..a9ca4b1354 100644
--- a/subsys/usb/class/usb_dfu.c
+++ b/subsys/usb/class/usb_dfu.c
@@ -679,10 +679,12 @@ static int dfu_custom_handle_req(struct usb_setup_packet *pSetup,
 				dfu_data.flash_area_id =
 				    FLASH_AREA_ID(image_0);
 				break;
+#if FLASH_AREA_LABEL_EXISTS(image_1)
 			case 1:
 				dfu_data.flash_area_id =
 				    UPLOAD_FLASH_AREA_ID;
 				break;
+#endif
 			default:
 				LOG_WRN("Invalid DFU alternate setting");
 				return -ENOTSUP;
```

nvlsianpu replied:
applied

nvlsianpu committed:
usb/class/usb_dfu: makes USB_DFU_WAIT_DELAY_MS configurable

Kconfig USB_DFU_WAIT_DELAY_MS was hidden.
Added prompt for it so it can be configured.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

emob-nordic commented on west.yml:
```diff
@@ -93,8 +93,9 @@ manifest:
       path: modules/crypto/mbedtls
     - name: mcuboot
-      revision: e64c5f00d67cf858d1fdead7135f5d61de2a4c68
+      revision: pull/851/head
```
I believe this must go in first before we merge that?

nvlsianpu replied:
was introduced already in other PR

emob-nordic commented on subsys/usb/class/Kconfig:
```diff
@@ -146,11 +146,12 @@ config USB_DFU_CLASS
-	int
+	int "wait_for_usb_dfu() timeout"
 	depends on USB_DFU_CLASS
 	default 12000
```
Any range there?

nvlsianpu replied:
ok <1, 120> seconds

eobalski reviewed (commented):
- [x] 648ad18caa2be223e5948eac39fc1c33e7d454e0 should be dropped as this is not a part of the feature.

- [x] b73ee57c45f9afc6ebea7521106ba7be91a77e71 should be squashed into 4b4a05f96f29c5e863f101348373308cb6e91a30

- [x] 0b2c66365fdddc7dea6cc448d0456d2df1da8286 Typo in commit message: Us -> Use. Maybe: "Use MCUBoot implementation rather than Zephyr's."

- [x] f90874628c598b2c732581dd20992a45d1ca8fb5 Typo in commit message: manifes -> manifest.
      [edit] was dropped


- [x] I think this PR should be blocked until #29612 is merged first.

nvlsianpu commented:
@utzig @jfischer-phytec-iot @emob-nordic Please revisit

de-nordic reviewed (approved):
Looks OK.

Tested on nrf52840dk, with various basic samples, and worked reliably.

carlescufi commented:
DNM because #28459 is still not merged, and I am not convinced about it. See my comments there.

nvlsianpu commented:
@carlescufi #28459 is not required for this PR at all.

de-nordic commented:
> 
> 
> DNM because #28459 is still not merged, and I am not convinced about it. See my comments there.

@carlescufi  The #28459 is cosmetic and we can deal with it later. This PR addresses feature request, and coincidentally a bug within compilation, reported by user here https://github.com/zephyrproject-rtos/zephyr/issues/28473.

de-nordic commented:
@bpbradley I think this addresses the issue https://github.com/zephyrproject-rtos/zephyr/issues/28473, that has been reported by you.

eobalski reviewed (approved):
Tested. Works fine.

bpbradley commented:
> @bpbradley I think this addresses the issue #28473, that has been reported by you, please join us in the review.

Yes this PR works! Thanks for the ping.

nvlsianpu commented:
This PR introduces #ifdefry for making the compilation of zephyr `dfu/boot/mcuboot` module possible while using in MCUboot bootloader. In the future (not in this PR) I plan to rework this using https://github.com/zephyrproject-rtos/zephyr/issues/29963

carlescufi commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -96,6 +98,7 @@ struct mcuboot_v1_raw_header {
 				  BOOT_MAX_ALIGN)
 #define MAGIC_OFFS(bank_area) ((bank_area)->fa_size - BOOT_MAGIC_SZ)
 
+#ifndef CONFIG_MCUBOOT
```
Why not then use:
```
#ifdef CONFIG_BOOTLOADER_MCUBOOT
```
so that we don't depend on a Kconfig option that is outside the Zephyr tree?
